lib/spack/spack/repository.py

```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
         self._all_package_names = None
         self._provider_index = None
 
-        # If repo_dirs is empty, just use the configuration
-        if not repo_dirs:
-            import spack.config
-            repo_dirs = spack.config.get_config('repos')
-            if not repo_dirs:
-                raise NoRepoConfiguredError(
-                    "Spack configuration contains no package repositories.")
-
         # Add each repo to this path.
         for root in repo_dirs:
-            try:
-                repo = Repo(root, self.super_namespace)
-                self.put_last(repo)
-            except RepoError as e:
-                tty.warn("Failed to initialize repository at '%s'." % root,
-                         e.message,
-                         "To remove the bad repository, run this command:",
-                         "    spack repo rm %s" % root)
+            self.append_from_path(root)
+
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```

Comment:
> In this case, instead, enforcing a contract that makes a class behave like a built-in structure of the language results in code that is easier to modify, because you know what to expect.

I disagree there. As a user of lists I know what to expect. As someone mucking around inside the internals of the implementation of ```MutableSequence``` I don't know what to expect. Will I accidentally override a method? Will I forget to call a ```_private_method```? Are there data structures inside the implementation of ```MutableSequence``` which need to be maintained? If I want this to behave correctly as a ```MutableSequence``` then I have to be familiar with its implementation.

> as soon as we implement the prescribed abstract functions with the same semantic as list, every other method that is part of the protocol is already implemented and works "as expected", i.e. as a list.

The most important function from a user perspective for ```RepoPath``` is ```get```. Beyond that, adding additional behaviors to a class that doesn't need them may be confusing. Will a user want to do ```repo_path[4:9]```? What does it mean in terms of the outcome of ```RepoPath.get``` if I say ```repo_path.insert(3, x)```?

Also, knowing that ```RepoPath``` is a mutable sequence won't tell a user for example whether ```repo_path[0]``` or ```repo_path[-1]``` is the most important repo: MutableSequence doesn't have anything to say about order and although a user could guess, they are advised to understand the interface. ```RepoPath.first_repo``` is more suggestive of priority.

> With a class that resembles a list sometimes, but does not other times, following your reasoning we'll need to inspect the semantic of every function before modifying one.

I think this is generally true: when you add functionality to a class you need to understand its internal implementation. The exception is when you are using composition. That requirement still exists when inheriting from ```MutableSequence```

> probably we'll need to touch repos to solve #3115

I read through that and didn't get why repos were the issue there. Assuming they are, I'd prefer these sorts of changes be integrated into a PR directed at fixing that in order to demonstrate the usefulness of those changes.